[Serialization] Use full target architectures for swiftmodule files

jrose-apple opened the pull request:
Previously these used the same "major architecture" names that the `#if os(...)` query accepts, but that can be a problem when building for, say, both armv7 and armv7s, even if the content is "the same".

For a transition period where external build tools are involved, the compiler will look for "arm.swiftmodule" if it fails to find "armv7.swiftmodule" or any other 32-bit ARM architecture. No other Apple platform architectures are affected, and AFAIK no one's using the architecture-based layout on Linux or any other platforms.

rdar://problem/45174692

jrose-apple commented:
@swift-ci Please test

rjmccall reviewed (approved):
Seems reasonable.

jrose-apple commented:
@swift-ci Please clean test Linux

compnerd commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -142,6 +142,23 @@ bool SerializedModuleLoader::maybeDiagnoseArchitectureMismatch(
 }
 
+static std::pair<llvm::SmallString<16>, llvm::SmallString<16>>
+getArchSpecificModuleFileNames(StringRef archName) {
```
Why not pass along the Triple rather than StringRef?  It feels like it would be nicer to derive that from the triple rather than the architecture name specifically.  We could have a boolean flag for whether it should do the fallback or not.

jrose-apple replied:
I don't think that actually makes the call sites nicer, and as a helper function that's the important thing.

compnerd commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -157,19 +174,19 @@ SerializedModuleLoaderBase::findModule(AccessPathElem moduleID,
-    archDocFile += file_types::getExtension(file_types::TY_SwiftModuleDocFile);
-  }
+  // FIXME: We used to use "major architecture" names for these files---the
+  // names checked in "#if os(...)". Fall back to that name in the one case
```
Did you really mean `#if os(...)` or do you mean `#if arch(...)`?

jrose-apple replied:
Oops, thanks!

jrose-apple committed:
[Serialization] Use full target architectures for swiftmodule files

Previously these used the same "major architecture" names that the
`#if os(...)` query accepts, but that can be a problem when building
for, say, both armv7 and armv7s, even if the content is "the same".

For a transition period where external build tools are involved, the
compiler will look for "arm.swiftmodule" if it fails to find
"armv7.swiftmodule" or any other 32-bit ARM architecture. No other
Apple platform architectures are affected, and AFAIK no one's using
the architecture-based layout on Linux or any other platforms.

rdar://problem/45174692

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please test source compatibility
